[v0.27.0] Skaffold fix panic

### Expected behavior
Skaffold fix should update skaffold.yaml file

### Actual behavior
Panic on run

### Information

- Skaffold version: v0.27.0
- Operating system: MacOS 10.14.4
- Contents of skaffold.yaml:

```yaml
apiVersion: skaffold/v1beta7
kind: Config
build:
  artifacts:
    - image: app
      context: .
      docker:
        dockerfile: ops/docker/app/Dockerfile
deploy:
  kubectl:
    manifests:
      - placeholder
      - ops/kubernetes/*.yaml
profiles:
  - name: local
    patches:
      - op: add
        path: /build/artifacts/0/docker
        value:
          buildArgs:
            PHP_INI: php.ini-development
      - op: add
        path: /build
        value:
          tagPolicy:
            sha256: {}
      - op: replace
        path: /deploy/kubectl/manifests/0
        value: ops/kubernetes/local/*.yaml
  - name: cluster
    patches:
      - op: add
        path: /build
        value:
          tagPolicy:
            gitCommit: {}
      - op: add
        path: /build
        value:
          googleCloudBuild:
            projectId: projectId
      - op: replace
        path: /deploy/kubectl/manifests/0
        value: ops/kubernetes/cluster/*.yaml
```

### Steps to reproduce the behavior

1. `skaffold fix`

```
WARN[0000] config version (skaffold/v1beta7) out of date: upgrading to latest (skaffold/v1beta8)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x14ebf83]

goroutine 1 [running]:
github.com/GoogleContainerTools/skaffold/vendor/gopkg.in/yaml%2ev2.handleErr(0xc0002bdc68)
        vendor/gopkg.in/yaml.v2/yaml.go:249 +0x9a
panic(0x218aa80, 0x34010c0)
        /usr/local/Cellar/go/1.12.4/libexec/src/runtime/panic.go:522 +0x1b5
github.com/GoogleContainerTools/skaffold/vendor/github.com/krishicks/yaml-patch.(*Node).MarshalYAML(0xc000595a20, 0x22273a0, 0xc000595a20, 0x3f82d50, 0xc000595a20)
        vendor/github.com/krishicks/yaml-patch/node.go:26 +0x33
github.com/GoogleContainerTools/skaffold/vendor/gopkg.in/yaml%2ev2.(*encoder).marshal(0xc0000de840, 0x0, 0x0, 0x22273a0, 0xc000572110, 0x196)
        vendor/gopkg.in/yaml.v2/encode.go:98 +0x82a
github.com/GoogleContainerTools/skaffold/vendor/gopkg.in/yaml%2ev2.(*encoder).structv.func1()
        vendor/gopkg.in/yaml.v2/encode.go:198 +0x55a
github.com/GoogleContainerTools/skaffold/vendor/gopkg.in/yaml%2ev2.(*encoder).mappingv(0xc0000de840, 0x0, 0x0, 0xc0002bcb80)
        vendor/gopkg.in/yaml.v2/encode.go:228 +0x147
github.com/GoogleContainerTools/skaffold/vendor/gopkg.in/yaml%2ev2.(*encoder).structv(0xc0000de840, 0x0, 0x0, 0x2267da0, 0xc0005720e0, 0x199)
        vendor/gopkg.in/yaml.v2/encode.go:185 +0xf7
github.com/GoogleContainerTools/skaffold/vendor/gopkg.in/yaml%2ev2.(*encoder).marshal(0xc0000de840, 0x0, 0x0, 0x2267da0, 0xc0005720e0, 0x199)
        vendor/gopkg.in/yaml.v2/encode.go:132 +0x726
github.com/GoogleContainerTools/skaffold/vendor/gopkg.in/yaml%2ev2.(*encoder).slicev(0xc0000de840, 0x0, 0x0, 0x20c2ac0, 0xc0000b4b30, 0x197)
        vendor/gopkg.in/yaml.v2/encode.go:244 +0x1f8
github.com/GoogleContainerTools/skaffold/vendor/gopkg.in/yaml%2ev2.(*encoder).marshal(0xc0000de840, 0x0, 0x0, 0x20c2ac0, 0xc0000b4b30, 0x197)
        vendor/gopkg.in/yaml.v2/encode.go:138 +0x412
github.com/GoogleContainerTools/skaffold/vendor/gopkg.in/yaml%2ev2.(*encoder).structv.func1()
        vendor/gopkg.in/yaml.v2/encode.go:198 +0x55a
github.com/GoogleContainerTools/skaffold/vendor/gopkg.in/yaml%2ev2.(*encoder).mappingv(0xc0000de840, 0x0, 0x0, 0xc0002bd108)
        vendor/gopkg.in/yaml.v2/encode.go:228 +0x147
github.com/GoogleContainerTools/skaffold/vendor/gopkg.in/yaml%2ev2.(*encoder).structv(0xc0000de840, 0x0, 0x0, 0x2267e60, 0xc0000b4a80, 0x199)
        vendor/gopkg.in/yaml.v2/encode.go:185 +0xf7
github.com/GoogleContainerTools/skaffold/vendor/gopkg.in/yaml%2ev2.(*encoder).marshal(0xc0000de840, 0x0, 0x0, 0x2267e60, 0xc0000b4a80, 0x199)
        vendor/gopkg.in/yaml.v2/encode.go:132 +0x726
github.com/GoogleContainerTools/skaffold/vendor/gopkg.in/yaml%2ev2.(*encoder).slicev(0xc0000de840, 0x0, 0x0, 0x20c2b00, 0xc000572360, 0x197)
        vendor/gopkg.in/yaml.v2/encode.go:244 +0x1f8
github.com/GoogleContainerTools/skaffold/vendor/gopkg.in/yaml%2ev2.(*encoder).marshal(0xc0000de840, 0x0, 0x0, 0x20c2b00, 0xc000572360, 0x197)
        vendor/gopkg.in/yaml.v2/encode.go:138 +0x412
github.com/GoogleContainerTools/skaffold/vendor/gopkg.in/yaml%2ev2.(*encoder).structv.func1()
        vendor/gopkg.in/yaml.v2/encode.go:198 +0x55a
github.com/GoogleContainerTools/skaffold/vendor/gopkg.in/yaml%2ev2.(*encoder).mappingv(0xc0000de840, 0x0, 0x0, 0xc0002bd690)
        vendor/gopkg.in/yaml.v2/encode.go:228 +0x147
github.com/GoogleContainerTools/skaffold/vendor/gopkg.in/yaml%2ev2.(*encoder).structv(0xc0000de840, 0x0, 0x0, 0x2267f20, 0xc0005722a0, 0x199)
        vendor/gopkg.in/yaml.v2/encode.go:185 +0xf7
github.com/GoogleContainerTools/skaffold/vendor/gopkg.in/yaml%2ev2.(*encoder).marshal(0xc0000de840, 0x0, 0x0, 0x2267f20, 0xc0005722a0, 0x199)
        vendor/gopkg.in/yaml.v2/encode.go:132 +0x726
github.com/GoogleContainerTools/skaffold/vendor/gopkg.in/yaml%2ev2.(*encoder).marshal(0xc0000de840, 0x0, 0x0, 0x21921e0, 0xc0005722a0, 0x16)
        vendor/gopkg.in/yaml.v2/encode.go:126 +0x59f
github.com/GoogleContainerTools/skaffold/vendor/gopkg.in/yaml%2ev2.(*encoder).marshalDoc(0xc0000de840, 0x0, 0x0, 0x21921e0, 0xc0005722a0, 0x16)
        vendor/gopkg.in/yaml.v2/encode.go:80 +0x10f
github.com/GoogleContainerTools/skaffold/vendor/gopkg.in/yaml%2ev2.Marshal(0x21921e0, 0xc0005722a0, 0x0, 0x0, 0x0, 0x0, 0x0)
        vendor/gopkg.in/yaml.v2/yaml.go:203 +0x329
github.com/GoogleContainerTools/skaffold/cmd/skaffold/app/cmd.runFix(0x25b7220, 0xc0000bc000, 0x23878f8, 0xd, 0xc00050fc00, 0x0, 0x0)
        cmd/skaffold/app/cmd/fix.go:61 +0x10e
github.com/GoogleContainerTools/skaffold/cmd/skaffold/app/cmd.NewCmdFix.func1(0xc000204500, 0x3446000, 0x0, 0x0, 0x0, 0x0)
        cmd/skaffold/app/cmd/fix.go:36 +0x56
github.com/GoogleContainerTools/skaffold/vendor/github.com/spf13/cobra.(*Command).execute(0xc000204500, 0x3446000, 0x0, 0x0, 0xc000204500, 0x3446000)
        vendor/github.com/spf13/cobra/command.go:762 +0x465
github.com/GoogleContainerTools/skaffold/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc00019cc80, 0xc0000bc000, 0x25b7220, 0xc0000c0000)
        vendor/github.com/spf13/cobra/command.go:852 +0x2ec
github.com/GoogleContainerTools/skaffold/vendor/github.com/spf13/cobra.(*Command).Execute(...)
        vendor/github.com/spf13/cobra/command.go:800
github.com/GoogleContainerTools/skaffold/cmd/skaffold/app.Run(0x25b7220, 0xc0000bc000, 0x25b7220, 0xc0000c0000, 0xc00050ff88, 0x100554f)
        cmd/skaffold/app/skaffold.go:35 +0xd3
main.main()
        cmd/skaffold/skaffold.go:30 +0x4e
```

--- a/pkg/skaffold/schema/latest/config.go
+++ b/pkg/skaffold/schema/latest/config.go
@@ -18,7 +18,6 @@ package latest
 
 import (
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/util"
-	yamlpatch "github.com/krishicks/yaml-patch"
 )
 
 const Version string = "skaffold/v1beta9"
@@ -539,7 +538,7 @@ type JSONPatch struct {
 	From string `yaml:"from,omitempty"`
 
 	// Value is the value to apply. Can be any portion of yaml.
-	Value *yamlpatch.Node `yaml:"value,omitempty"`
+	Value *util.YamlpatchNode `yaml:"value,omitempty"`
 }
 
 // Activation criteria by which a profile is auto-activated.

--- a/pkg/skaffold/schema/profiles.go
+++ b/pkg/skaffold/schema/profiles.go
@@ -159,11 +159,16 @@ func applyProfile(config *latest.SkaffoldConfig, profile latest.Profile) error {
 			op = "replace"
 		}
 
+		var value *yamlpatch.Node
+		if v := patch.Value; v != nil {
+			value = &v.Node
+		}
+
 		patch := yamlpatch.Operation{
 			Op:    yamlpatch.Op(op),
 			Path:  yamlpatch.OpPath(patch.Path),
 			From:  yamlpatch.OpPath(patch.From),
-			Value: patch.Value,
+			Value: value,
 		}
 
 		if !tryPatch(patch, buf) {

--- a/pkg/skaffold/schema/profiles_test.go
+++ b/pkg/skaffold/schema/profiles_test.go
@@ -24,6 +24,7 @@ import (
 	cfg "github.com/GoogleContainerTools/skaffold/pkg/skaffold/config"
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/constants"
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/latest"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/util"
 	"github.com/GoogleContainerTools/skaffold/testutil"
 	yamlpatch "github.com/krishicks/yaml-patch"
 	"k8s.io/client-go/tools/clientcmd/api"
@@ -248,7 +249,7 @@ func TestApplyProfiles(t *testing.T) {
 					Name: "profile",
 					Patches: []latest.JSONPatch{{
 						Path:  "/build/artifacts/0/docker/dockerfile",
-						Value: yamlpatch.NewNode(str("Dockerfile.DEV")),
+						Value: &util.YamlpatchNode{Node: *yamlpatch.NewNode(str("Dockerfile.DEV"))},
 					}},
 				}),
 			),

--- a/pkg/skaffold/schema/util/util.go
+++ b/pkg/skaffold/schema/util/util.go
@@ -21,6 +21,7 @@ import (
 	"reflect"
 	"strings"
 
+	yamlpatch "github.com/krishicks/yaml-patch"
 	"gopkg.in/yaml.v2"
 )
 
@@ -34,24 +35,64 @@ type HelmOverrides struct {
 	Values map[string]interface{} `yaml:",inline"`
 }
 
-type inlineYaml struct {
-	Yaml string
+// MarshalJSON implements JSON marshalling by including the value as an inline yaml fragment.
+func (h *HelmOverrides) MarshalJSON() ([]byte, error) {
+	return marshalInlineYaml(h)
 }
 
-func (h *HelmOverrides) MarshalJSON() ([]byte, error) {
-	yaml, err := yaml.Marshal(h)
+// UnmarshalYAML implements JSON unmarshalling by reading an inline yaml fragment.
+func (h *HelmOverrides) UnmarshalJSON(text []byte) error {
+	yml, err := unmarshalInlineYaml(text)
+	if err != nil {
+		return err
+	}
+	return yaml.Unmarshal([]byte(yml), h)
+}
+
+// YamlpatchNode wraps a `yamlpatch.Node` and makes it serializable to JSON.
+// The yaml serialization needs to be implemented manually, because the node may be
+// an arbitrary yaml fragment so that a field tag `yaml:",inline"` does not work here.
+type YamlpatchNode struct {
+	// node is an arbitrary yaml fragment
+	Node yamlpatch.Node
+}
+
+// MarshalJSON implements JSON marshalling by including the value as an inline yaml fragment.
+func (n *YamlpatchNode) MarshalJSON() ([]byte, error) {
+	return marshalInlineYaml(n)
+}
+
+// UnmarshalYAML implements JSON unmarshalling by reading an inline yaml fragment.
+func (n *YamlpatchNode) UnmarshalJSON(text []byte) error {
+	yml, err := unmarshalInlineYaml(text)
+	if err != nil {
+		return err
+	}
+	return yaml.Unmarshal([]byte(yml), n)
+}
+
+// MarshalYAML implements yaml.Marshaler.
+func (n *YamlpatchNode) MarshalYAML() (interface{}, error) {
+	return n.Node.MarshalYAML()
+}
+
+// UnmarshalYAML implements yaml.Unmarshaler
+func (n *YamlpatchNode) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	return n.Node.UnmarshalYAML(unmarshal)
+}
+
+func marshalInlineYaml(in interface{}) ([]byte, error) {
+	yaml, err := yaml.Marshal(in)
 	if err != nil {
 		return nil, err
 	}
-	return json.Marshal(inlineYaml{string(yaml)})
+	return json.Marshal(string(yaml))
 }
 
-func (h *HelmOverrides) UnmarshalJSON(text []byte) error {
-	in := inlineYaml{}
-	if err := json.Unmarshal(text, &in); err != nil {
-		return err
-	}
-	return yaml.Unmarshal([]byte(in.Yaml), h)
+func unmarshalInlineYaml(text []byte) (string, error) {
+	var in string
+	err := json.Unmarshal(text, &in)
+	return in, err
 }
 
 // IsOneOfField checks if a field is tagged with oneOf

--- a/pkg/skaffold/schema/util/util_test.go
+++ b/pkg/skaffold/schema/util/util_test.go
@@ -24,14 +24,14 @@ import (
 	yaml "gopkg.in/yaml.v2"
 )
 
-const overridesYaml string = `global:
+const yamlFragment string = `global:
   enabled: true
   localstack: {}
 `
 
 func TestHelmOverridesMarshalling(t *testing.T) {
 	h := &HelmOverrides{}
-	err := yaml.Unmarshal([]byte(overridesYaml), h)
+	err := yaml.Unmarshal([]byte(yamlFragment), h)
 	testutil.CheckError(t, false, err)
 
 	asJSON, err := json.Marshal(h)
@@ -41,12 +41,12 @@ func TestHelmOverridesMarshalling(t *testing.T) {
 	testutil.CheckError(t, false, err)
 
 	actual, err := yaml.Marshal(h)
-	testutil.CheckErrorAndDeepEqual(t, false, err, overridesYaml, string(actual))
+	testutil.CheckErrorAndDeepEqual(t, false, err, yamlFragment, string(actual))
 }
 
 func TestHelmOverridesWhenEmbedded(t *testing.T) {
 	h := HelmOverrides{}
-	err := yaml.Unmarshal([]byte(overridesYaml), &h)
+	err := yaml.Unmarshal([]byte(yamlFragment), &h)
 	testutil.CheckError(t, false, err)
 
 	out, err := yaml.Marshal(struct {
@@ -54,6 +54,37 @@ func TestHelmOverridesWhenEmbedded(t *testing.T) {
 	}{h})
 
 	testutil.CheckErrorAndDeepEqual(t, false, err, `overrides:
+  global:
+    enabled: true
+    localstack: {}
+`, string(out))
+}
+
+func TestYamlpatchNodeMarshalling(t *testing.T) {
+	n := &YamlpatchNode{}
+	err := yaml.Unmarshal([]byte(yamlFragment), n)
+	testutil.CheckError(t, false, err)
+
+	asJSON, err := json.Marshal(n)
+	testutil.CheckError(t, false, err)
+
+	err = json.Unmarshal(asJSON, n)
+	testutil.CheckError(t, false, err)
+
+	actual, err := yaml.Marshal(n)
+	testutil.CheckErrorAndDeepEqual(t, false, err, yamlFragment, string(actual))
+}
+
+func TestYamlpatchNodeWhenEmbedded(t *testing.T) {
+	n := &YamlpatchNode{}
+	err := yaml.Unmarshal([]byte(yamlFragment), &n)
+	testutil.CheckError(t, false, err)
+
+	out, err := yaml.Marshal(struct {
+		Node *YamlpatchNode `yaml:"value,omitempty"`
+	}{n})
+
+	testutil.CheckErrorAndDeepEqual(t, false, err, `value:
   global:
     enabled: true
     localstack: {}

--- a/pkg/skaffold/schema/v1beta5/config.go
+++ b/pkg/skaffold/schema/v1beta5/config.go
@@ -18,7 +18,6 @@ package v1beta5
 
 import (
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/util"
-	yamlpatch "github.com/krishicks/yaml-patch"
 )
 
 const Version string = "skaffold/v1beta5"
@@ -499,7 +498,7 @@ type JSONPatch struct {
 	From string `yaml:"from,omitempty"`
 
 	// Value is the value to apply. Can be any portion of yaml.
-	Value *yamlpatch.Node `yaml:"value,omitempty"`
+	Value *util.YamlpatchNode `yaml:"value,omitempty"`
 }
 
 // Activation criteria by which a profile is auto-activated.

--- a/pkg/skaffold/schema/v1beta6/config.go
+++ b/pkg/skaffold/schema/v1beta6/config.go
@@ -18,7 +18,6 @@ package v1beta6
 
 import (
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/util"
-	yamlpatch "github.com/krishicks/yaml-patch"
 )
 
 const Version string = "skaffold/v1beta6"
@@ -523,7 +522,7 @@ type JSONPatch struct {
 	From string `yaml:"from,omitempty"`
 
 	// Value is the value to apply. Can be any portion of yaml.
-	Value *yamlpatch.Node `yaml:"value,omitempty"`
+	Value *util.YamlpatchNode `yaml:"value,omitempty"`
 }
 
 // Activation criteria by which a profile is auto-activated.

--- a/pkg/skaffold/schema/v1beta7/config.go
+++ b/pkg/skaffold/schema/v1beta7/config.go
@@ -18,7 +18,6 @@ package v1beta7
 
 import (
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/util"
-	yamlpatch "github.com/krishicks/yaml-patch"
 )
 
 const Version string = "skaffold/v1beta7"
@@ -512,7 +511,7 @@ type JSONPatch struct {
 	From string `yaml:"from,omitempty"`
 
 	// Value is the value to apply. Can be any portion of yaml.
-	Value *yamlpatch.Node `yaml:"value,omitempty"`
+	Value *util.YamlpatchNode `yaml:"value,omitempty"`
 }
 
 // Activation criteria by which a profile is auto-activated.

--- a/pkg/skaffold/schema/v1beta8/config.go
+++ b/pkg/skaffold/schema/v1beta8/config.go
@@ -18,7 +18,6 @@ package v1beta8
 
 import (
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/util"
-	yamlpatch "github.com/krishicks/yaml-patch"
 )
 
 const Version string = "skaffold/v1beta8"
@@ -539,7 +538,7 @@ type JSONPatch struct {
 	From string `yaml:"from,omitempty"`
 
 	// Value is the value to apply. Can be any portion of yaml.
-	Value *yamlpatch.Node `yaml:"value,omitempty"`
+	Value *util.YamlpatchNode `yaml:"value,omitempty"`
 }
 
 // Activation criteria by which a profile is auto-activated.

--- a/pkg/skaffold/schema/validation/validation.go
+++ b/pkg/skaffold/schema/validation/validation.go
@@ -54,6 +54,9 @@ func visitStructs(s interface{}, visitor func(interface{}) error) []error {
 
 		// also check all fields of the current struct
 		for i := 0; i < t.NumField(); i++ {
+			if !v.Field(i).CanInterface() {
+				continue
+			}
 			if fieldErrs := visitStructs(v.Field(i).Interface(), visitor); fieldErrs != nil {
 				errs = append(errs, fieldErrs...)
 			}

--- a/pkg/skaffold/schema/validation/validation_test.go
+++ b/pkg/skaffold/schema/validation/validation_test.go
@@ -192,6 +192,47 @@ func TestVisitStructs(t *testing.T) {
 			},
 			expectedErrs: 2,
 		},
+		{
+			name: "unexported fields",
+			input: struct {
+				a emptyStruct
+			}{},
+			expectedErrs: 1,
+		},
+		{
+			name: "exported and unexported fields",
+			input: struct {
+				a, A, b emptyStruct
+			}{},
+			expectedErrs: 2,
+		},
+		{
+			name: "unexported nil ptr fields",
+			input: struct {
+				a *emptyStruct
+			}{},
+			expectedErrs: 1,
+		},
+		{
+			name: "unexported ptr fields",
+			input: struct {
+				a *emptyStruct
+			}{
+				a: &emptyStruct{},
+			},
+			expectedErrs: 1,
+		},
+		{
+			name: "unexported and exported ptr fields",
+			input: struct {
+				a, A, b *emptyStruct
+			}{
+				a: &emptyStruct{},
+				A: &emptyStruct{},
+				b: &emptyStruct{},
+			},
+			expectedErrs: 2,
+		},
 	}
 
 	for _, test := range tests {

--- a/pkg/skaffold/schema/validation/validation_test.go
+++ b/pkg/skaffold/schema/validation/validation_test.go
@@ -196,21 +196,29 @@ func TestVisitStructs(t *testing.T) {
 			name: "unexported fields",
 			input: struct {
 				a emptyStruct
-			}{},
+			}{
+				a: emptyStruct{},
+			},
 			expectedErrs: 1,
 		},
 		{
 			name: "exported and unexported fields",
 			input: struct {
 				a, A, b emptyStruct
-			}{},
+			}{
+				a: emptyStruct{},
+				A: emptyStruct{},
+				b: emptyStruct{},
+			},
 			expectedErrs: 2,
 		},
 		{
 			name: "unexported nil ptr fields",
 			input: struct {
 				a *emptyStruct
-			}{},
+			}{
+				a: nil,
+			},
 			expectedErrs: 1,
 		},
 		{

